src/github.com/matrix-org/dendrite/syncserver/storage/output_room_events_table.go

```diff
@@ -244,7 +267,14 @@ func rowsToEvents(rows *sql.Rows) ([]gomatrixserverlib.Event, error) {
 		if err != nil {
 			return nil, err
 		}
-		result = append(result, ev)
+		if order == rowOrderForward {
+			result = append(result, ev)
+		} else if order == rowOrderReverse {
+			result = append([]gomatrixserverlib.Event{ev}, result...)
```

Comment:
It's a consequence of how go represents slices internally. They are represented as a pointer to the start of the list, a length and a capacity. When you append to the end of a slice it only needs to reallocate if there isn't enough capacity. If it does need to reallocate then it'll need to copy the contents of the backing array somewhere. When it reallocates it'll probably double the size of the backing array. So it'll take O(n log(n)) operations to fill out a list by appending to the end. (If you pre-allocated the backing array with the correct length then it will be O(n) operations).

However if you "append" to the beginning then you have to copy the entire array for every insertion which will be O(n^2) operations.